src/libraries/System.Private.CoreLib/src/System/Text/Latin1Utility.Helpers.cs

```diff
@@ -0,0 +1,111 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Runtime.Intrinsics;
+using System.Runtime.Intrinsics.X86;
+using Internal.Runtime.CompilerServices;
+
+namespace System.Text
+{
+    internal static partial class Latin1Utility
+    {
+        /// <summary>
+        /// Returns <see langword="true"/> iff all chars in <paramref name="value"/> are Latin-1.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool AllCharsInUInt32AreLatin1(uint value)
+        {
+            return (value & ~0x00FF00FFu) == 0;
+        }
+
+        /// <summary>
+        /// Returns <see langword="true"/> iff all chars in <paramref name="value"/> are Latin-1.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool AllCharsInUInt64AreLatin1(ulong value)
+        {
+            return (value & ~0x00FF00FF_00FF00FFul) == 0;
+        }
+
+        /// <summary>
+        /// Given a DWORD which represents two packed chars in machine-endian order,
+        /// <see langword="true"/> iff the first char (in machine-endian order) is Latin-1.
+        /// </summary>
+        /// <param name="value"></param>
+        /// <returns></returns>
+        private static bool FirstCharInUInt32IsLatin1(uint value)
+        {
+            return (BitConverter.IsLittleEndian && (value & 0xFF00u) == 0)
+                || (!BitConverter.IsLittleEndian && (value & 0xFF000000u) == 0);
+        }
+
+        /// <summary>
+        /// Given a QWORD which represents a buffer of 4 Latin-1 chars in machine-endian order,
+        /// narrows each WORD to a BYTE, then writes the 4-byte result to the output buffer
+        /// also in machine-endian order.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void NarrowFourUtf16CharsToLatin1AndWriteToBuffer(ref byte outputBuffer, ulong value)
+        {
+            Debug.Assert(AllCharsInUInt64AreLatin1(value));
+
+            if (Sse2.X64.IsSupported)
+            {
+                // Narrows a vector of words [ w0 w1 w2 w3 ] to a vector of bytes
+                // [ b0 b1 b2 b3 b0 b1 b2 b3 ], then writes 4 bytes (32 bits) to the destination.
+
+                Vector128<short> vecWide = Sse2.X64.ConvertScalarToVector128UInt64(value).AsInt16();
```

Comment:
We also have `LoadScalarVector128(ulong*)` for 32-bit and I believe `value` is unlikely to be split between 2 registers.

Reply:
By the time this method is called we already have a `ulong` (as a value), not a `ulong*`. I tried using `&value` hoping that the JIT would optimize it since the method is marked as aggressive inlining, but it still results in a stack spill.

Is there a similar method that takes a `ulong` directly as a value? For context, this method is currently hit in x64 and arm64 code paths, but not x86 or arm32. Though there's nothing that would inherently prevent it from working if called by a 32-bit process.